scripts/migrations/2024_11_move_obf_opf_opff_products_to_off_dirs.pl

```diff
@@ -0,0 +1,258 @@
+#!/usr/bin/perl -w
+
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2019 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use CGI::Carp qw(fatalsToBrowser);
+
+use ProductOpener::PerlStandards;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Store qw/:all/;
+use ProductOpener::Index qw/:all/;
+use ProductOpener::Display qw/:all/;
+use ProductOpener::Tags qw/:all/;
+use ProductOpener::Users qw/:all/;
+use ProductOpener::Images qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Mail qw/:all/;
+use ProductOpener::Products qw/:all/;
+use ProductOpener::Food qw/:all/;
+use ProductOpener::Ingredients qw/:all/;
+use ProductOpener::Images qw/:all/;
+use ProductOpener::Data qw/:all/;
+use ProductOpener::Orgs qw/:all/;
+use ProductOpener::Paths qw/:all/;
+
+use CGI qw/:cgi :form escapeHTML/;
+use URI::Escape::XS;
+use Storable qw/dclone/;
+use Encode;
+use JSON::PP;
+
+use File::Copy::Recursive qw/dirmove/;
+
+use Data::Dumper;
+
+# Get a list of all products
+
+use Getopt::Long;
+
+my $move = 0;
+
+GetOptions('move' => \$move,);
+
+my @dirs_without_product_sto = ();
+my @empty_dirs = ();
+my @products_existing_on_off = ();
+my @products_not_existing_on_off = ();
+my @products_existing_on_off_but_deleted_on_off = ();
+my @products_existing_on_off_but_deleted_locally = ();
+my @products_existing_on_off_and_not_deleted_on_off_or_locally = ();
+
+# We removed the other servers dirs from the base paths,
+# but for this script, we need /srv/off/products and /srv/off/html/images/products
+
+$BASE_DIRS{OFF_PRODUCTS} = "/srv/off/products";
+$BASE_DIRS{OFF_PRODUCTS_IMAGES} = "/srv/off/html/images/products";
+
+sub move_product_dir_to_off ($dir, $dir2, $dir3, $dir4) {
+	# move .sto files
+	print STDERR "moving $dir/$dir2/$dir3/$dir4\n";
```

Comment:
For me, it's incredible that in this whole function you never put $dir/$dir2/$dir3/$dir4 in a variable to avoid copying it everywhere around !

Reply:
It's just a one off script that will never be used again, so I mainly wanted to get this thing done as quickly as possible.